File: src/p2/Fraction.java
package src.p2;

public class Fraction {
    private int numerator;   // Чисельник
    private int denominator; // Знаменник

    // Конструктор
    public Fraction(int numerator, int denominator) {
        if (denominator == 0) {
            throw new ArithmeticException("Знаменник не може бути нулем!");
        }
        this.numerator = numerator;
        this.denominator = denominator;
        reductionFraction(); // Автоматично скорочуємо дріб при створенні
    }

    // Метод додавання
    public Fraction addition(Fraction other) {
        int newNumerator = this.numerator * other.denominator + other.numerator * this.denominator;
        int newDenominator = this.denominator * other.denominator;
        return new Fraction(newNumerator, newDenominator);
    }

    // Метод віднімання
    public Fraction subtraction(Fraction other) {
        int newNumerator = this.numerator * other.denominator - other.numerator * this.denominator;
        int newDenominator = this.denominator * other.denominator;
        return new Fraction(newNumerator, newDenominator);
    }

    // Метод множення
    public Fraction multiplication(Fraction other) {
        int newNumerator = this.numerator * other.numerator;
        int newDenominator = this.denominator * other.denominator;
        return new Fraction(newNumerator, newDenominator);
    }

    // Метод ділення
    public Fraction division(Fraction other) {
        if (other.numerator == 0) {
            throw new ArithmeticException("Ділення на нуль неможливе!");
        }
        int newNumerator = this.numerator * other.denominator;
        int newDenominator = this.denominator * other.numerator;
        return new Fraction(newNumerator, newDenominator);
    }

    // Метод скорочення дробу
    private void reductionFraction() {
        int gcd = gcd(numerator, denominator);
        numerator /= gcd;
        denominator /= gcd;
    }

    // Метод для знаходження найбільшого спільного дільника (НСД)
    private int gcd(int a, int b) {
        while (b != 0) {
            int temp = b;
            b = a % b;
            a = temp;
        }
        return Math.abs(a);
    }

    // Метод для виведення дробу у вигляді рядка
    @Override
    public String toString() {
        return numerator + "/" + denominator;
    }
}

File: fraction.py
import math

class Fraction:
    def __init__(self, numerator, denominator):
        if denominator == 0:
            raise ValueError("Denominator cannot be zero!")
        self.numerator = numerator
        self.denominator = denominator

    def addition(self, other):
        # (a/b + c/d) = (ad + bc)/bd
        num = self.numerator * other.denominator + other.numerator * self.denominator
        den = self.denominator * other.denominator
        return Fraction(num, den).reductionFraction()

    def subtraction(self, other):
        # (a/b - c/d) = (ad - bc)/bd
        num = self.numerator * other.denominator - other.numerator * self.denominator
        den = self.denominator * other.denominator
        return Fraction(num, den).reductionFraction()

    def multiplication(self, other):
        # (a/b * c/d) = (ac)/(bd)
        num = self.numerator * other.numerator
        den = self.denominator * other.denominator
        return Fraction(num, den).reductionFraction()

    def division(self, other):
        # (a/b) / (c/d) = (a*d)/(b*c)
        if other.numerator == 0:
            raise ZeroDivisionError("Cannot divide by a fraction with numerator 0!")
        num = self.numerator * other.denominator
        den = self.denominator * other.numerator
        return Fraction(num, den).reductionFraction()

    def reductionFraction(self):
        # Скорочення дробу
        gcd_value = math.gcd(self.numerator, self.denominator)
        # Якщо знаменник від'ємний – виправляємо знак
        if self.denominator < 0:
            gcd_value = -gcd_value
        return Fraction(self.numerator // gcd_value, self.denominator // gcd_value)

    def __str__(self):
        return f"{self.numerator}/{self.denominator}"
